tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1864,6 +1864,50 @@ void CGMSHLSLRuntime::AddHLSLFunctionInfo(Function *F, const FunctionDecl *FD) {
       funcProps->numThreads[1] = 1;
       funcProps->numThreads[2] = 1;
     }
+
+    // Validate Compute Shader system value inputs per launch mode
+    for (ParmVarDecl *param : FD->parameters()) {
```

Comment:
Checking this properly is more complicated.  You can supply a struct as input with these fields.  The struct could have these fields in nested struct or base class.  I think we need a more general function that walks entry function parameters, identifying the inputs and outputs that they declare, with their semantics, types and locations, for general use in entry function validation, rather than ad-hoc checks that don't catch all possible cases.

Reply:
I think the right approach would be to write a few CollectSignatureElements function overloads to solve this.

Here's a rough outline:

- Gathers into an info structure containing a list of elements for each SigPoint.  Elements have:
    - a component type, rows, and cols
    - the semantic name, and a starting semantic index.  Maybe we can capture the indexing pattern later, but don't need to yet.
    - interpolation modifier, and source location, if any
    - a source location for the parameter or field associated with this signature element
    - a source location for the associated semantic annotation (may not be on this field)
- First level:
    - input: entry function and ShaderKind.
    - Processes parameters, return type and function semantic.
- Second level recursive:
    - input: type, semantic name and index (if any), and interpolation modifiers (if any), array information (if any) - note that arrays complicate things, so details need to be worked out here.
    - Collects array dimensions, walks struct bases and fields with recursion
    - Outer semantic assignments override inner ones, with semantic indexing based on the flattening shown in [this example](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#signature-record-metadata "https://github.com/microsoft/directxshadercompiler/blob/main/docs/dxil.rst#signature-record-metadata").
- During traversal, diagnostics can be emitted when there's a clear error based on the hctdb-specified element definitions.
- After traversal, additional rules can be diagnosed from the collected elements, such as the launch mode specific constraints.

Reply:
Why are we doing this in CodeGen instead of Sema?

Reply:
Yes, this should be in Sema instead of CodeGen as well.  I think when I looked at this code and wrote the recommended approach, I was imagining it happening in Sema, since it was operating on the clang types.  It makes virtually no difference to what needs to be done, but Sema is the right place to do it.